src/web100srv.c

```diff
@@ -1527,647 +1382,422 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
-  // around.  If we don't we get zombies. The pkt-pair handling
-  // created 2 childern and we need to get rid of them. To know
-  // what PIDs to look for on 1/19/10 the run_test() routine was
-  // modified to pass around these values. The values are set in
-  // the proper routine in the testoptions.c file.
-  // Then add call to child_sig() routine and pass in these PID's
-  // so we handle each child in sequence
-
-  log_println(6, "remove pkt-pair children c2s=%d, s2c=%d", testopt.child1,
-              testopt.child2);
-  if (testopt.child1 != 0) child_sig(testopt.child1);
-  if (testopt.child2 != 0) child_sig(testopt.child2);
   exit(0);
 }
 
+static int max(int a, int b) { return (a > b) ? a : b; }
+
 /**
- * The server's main loop.  This is the function that, once all arguments are
- * processed and the server environment has been set up, will keep waiting for
- * new connections and then forking off children to handle those connections.
+ * Waits for a new client to connect, for one of the child processes to die, or
+ * (optionally) until a timeout occurs.  If the select() is woken up by a new
+ * client connecting, then this routine will return true.
+ * @param listenfd The file descriptor which listens to the network
+ * @param signalfd The pipe which will be written to upon receipt of SIGCHILD
+ * @param wait_forever True if the server should wait forever
+ * @return True if there is a new client on listenfd
  */
-void NDT_server_main_loop(SSL_CTX *ssl_context, int listenfd) {
-  int child_pipe[2];
-  int tpid, mwaiting = 0;
-  fd_set rfd;
+int wait_for_wakeup(int listenfd, int signalfd, int wait_forever) {
+  fd_set fds;
   int retcode;
-  time_t tt;
+  ServerWakeupMessage signal_value;
+  struct timeval sel_tv, *psel_tv = NULL;
+
+  FD_ZERO(&fds);
+  FD_SET(listenfd, &fds);
+  FD_SET(signalfd, &fds);
+  if (!wait_forever) {
+    sel_tv.tv_sec = 3;  // Timeout after 3 seconds
+    sel_tv.tv_usec = 0;
+    psel_tv = &sel_tv;
+  }
+  retcode = select(max(listenfd, signalfd) + 1, &fds, NULL, NULL, psel_tv);
+  if (retcode == -1) {
+    if (errno != EINTR) {
+      // EINTR is expected every now and then due to signal handling.
+      log_println(0, "Error in server's select call: %d (%s)", errno,
+                  strerror(errno));
+    }
+    return 0;
+  }
+  // If we woke up due to data on the signalfd, then clear the buffer.
+  if (FD_ISSET(signalfd, &fds)) {
+    read(signalfd, &signal_value, sizeof(ServerWakeupMessage));
+  }
+  return FD_ISSET(listenfd, &fds);
+}
+
+/**
+ * There is a new client waiting to start on the listenfd socket. Accept the
+ * connection, and fork off a new process to run all the associated tests.
+ * @param listenfd The file descriptor which listens to the network
+ * @param ssl_context The ssl_context for any new connections
+ * @return NULL or a newly initialized ndtchild struct - the calling function
+ *         owns the struct and its memory
```

Comment:
>>>> Does it only return NULL on error? Can we document this?

>>>It returns NULL if it was woken up by something other than a new client connecting. This is not necessarily an error (every test completing ends up sending SIGCHLD), it just means that there may be work to be done even if there isn't any new client connecting.

>>I'd still like to see clearer documentation. What does it mean to the caller if it gets back NULL?

>>Also, can we clarify the documentation for the function description overall? I'm still not clear on what this function does. I understand waiting for a child for infinity or until a specified timeout, but why wake up if there's no new child and we didn't time out?

>If they get NULL, then there was no new ndtchild but the server woke up anyway.

>The server would also wake up from its select() if it received a wake up message on signalfd. That would happen if a client finished their tests and closed the connection, making it so the server has queue space to run a new client. Previous versions of the server could only start queued clients after the server timed out waiting for new clients to connect. This uselessly added 3 seconds to every queued test and was tangled up with bad queue maintenance code. Now there is a signaling channel that gets used whenever a client exit()s, so that the server can be responsive when clients finish.

>That's the whole reason signalfd exists.

I don't think this is true any longer. `NULL` now seems to exclusively mean an error. Regardless, the docstring should define what `NULL` vs. non-`NULL` means to the caller.